histomicstk/segmentation/label/isbfcpp.cpp

```diff
@@ -0,0 +1,344 @@
+/*
+C++ version of ISBF for TraceBounds
+*/
+#include <iostream>
+#include <math.h>
+#include "isbfcpp.h"
+
+using namespace std;
+
+isbfcpp::isbfcpp()
+{
+    head = NULL;
+    curr = NULL;
+}
+
+int isbfcpp::length()
+{
+  node *cur = head;
+  int len = 0;
+	while(cur) {
+		cur = cur->next;
+    len++;
+	}
+  return len;
+}
+
+void isbfcpp::nth_from_last(int n, int &x, int &y)
+{
+    node *cur = head;
+    int count = 0;
+  	while(cur) {
+        if (count == n) {
+            x = cur->val_x;
+            y = cur->val_y;
+        }
+    		cur = cur->next;
+        count++;
+  	}
+}
+
+bool isbfcpp::addList(int x, int y)
+{
+  node *newNode = new node;
+  newNode->val_x = x;
+  newNode->val_y = y;
+  newNode->next = NULL;
+
+  if (head == NULL) {
+    head = new node;
+    curr = new node;
+    head = curr = newNode;
+  }
+  else {
+    curr-> next = newNode;
+    curr = newNode;
+  }
+
+  count++;
+  return true;
+}
+
+void isbfcpp::rotateMatrix(int rows, int cols, int **input, int **output)
+{
+    int i, j;
+    for (i=0; i<rows; i++){
+      for (j=0;j<cols; j++){
+        output[j][rows-1-i] = input[i][j];
+      }
+    }
+}
+
+void isbfcpp::clean()
+{
+    for (int i = 0; i < nrows; i++) {
+        delete[] matrix00[i];
+        delete[] matrix180[i];
+    }
+    for (int i = 0; i < ncols; i++) {
+        delete[] matrix90[i];
+        delete[] matrix270[i];
+    }
+    delete[] matrix00;
+    delete[] matrix90;
+    delete[] matrix180;
+    delete[] matrix270;
+}
+
+vector<int> isbfcpp::getList(int rows, int cols, int *size, int *mask, int startX, int startY, float inf)
+{
+    addList(startX, startY);
+
+    nrows = rows;
+    ncols = cols;
+
+    matrix00 = new int*[nrows];
+    for (int i = 0; i < nrows; i++) {
+      matrix00[i] = new int[ncols];
+    }
+
+    // copy mask to matrix00
+    for(int i=0; i< nrows; i++){
+        for(int j=0; j< ncols; j++){
+            matrix00[i][j] = mask[i*ncols+j];
+        }
+    }
```

Comment:
You can initialize an std::vector from c array like [this](http://stackoverflow.com/questions/2434196/how-to-initialize-stdvector-from-c-style-array). Generalize this to 2D.